Detect and cleanly exit from renderer crashes

flotwig opened the pull request:
<!-- 
Thanks for contributing!
Read our contribution guidelines here: 
https://github.com/cypress-io/cypress/blob/develop/.github/CONTRIBUTING.md 
-->

<!-- Example: "Closes #1234" -->

- Closes #6170 

### User facing changelog

- Cypress will now detect if your browser tab has crashed and will cleanly exit.

### Additional details

- previously, Cypress would hang open until the CI job timed out
- methods used to tell if the browser tab has crashed:
	- if the user is not `cy.visit`ing, and the socket disconnects, and there is not a new runner connection within 5 seconds, the browser tab crashed
	- if the user is `cy.visit`ing, the socket disconnects, and it does not reconnect within 10 seconds (generous)
	- [x] ~~this makes the code a lot more complex, maybe the rules should just be "if the runner disconnects, and it does not reconnect within 10 seconds, the browser tab crashed" - don't try to be smart about it~~
		- apparently socket.io can and will disconnect and reconnect, so there should be a delay in either case
		- since 10 seconds is kinda a long delay, maybe these two cases should remain separate
- what do we do when an unexpected exit does occur?
	- browser is left open for debugging
	- in open mode: a warning is logged to stdio
	- in run mode: an error is thrown and tests end
- [x] if the extension d/c's in run mode, force exit
- [x] keep extension d/c error message the same

Most of the meat of the error has been moved to the docs. Docs PR: cypress-io/cypress-documentation#2413

## User Experience Changes

- users will now see the `BROWSER_CRASHED` error message in stdout if the browser crashes in open mode
- users will see their tests exit early if the browser crashes, or if the extension disconnects, in run mode
- webextension disconnections now cause tests to end in run mode

### PR Tasks

<!-- 
These tasks must be completed before a PR is merged.
Delete tasks if they are not applicable. 
-->

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub? <!-- (internal team only)-->
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? cypress-io/cypress-documentation#2413

flotwig committed:
tsify socket.js

flotwig committed:
add verbose debug logs for socket

flotwig committed:
remove reporter:connected as it is unused

flotwig committed:
remove more stuff related to unused reporter room

flotwig commented:
WIP, using the existing RENDERER_CRASHED error and cleaning it up a bit to be generic:

![image](https://user-images.githubusercontent.com/1151760/72648814-00ae2f80-394a-11ea-87f2-408db56c6c19.png)

In run mode, it will just throw and exit.

flotwig committed:
add code to handle unexpected d/c's

flotwig committed:
reduce timeout to 10s

flotwig committed:
update error

flotwig committed:
use runStatePreservedAt as navigation predictor

flotwig commented on packages/server/lib/errors.coffee:
```diff
@@ -528,23 +528,11 @@ getMsgByType = (type, arg1 = {}, arg2) ->
     when "RENDERER_CRASHED"
       """
-      We detected that the Chromium Renderer process just crashed.
+      We detected that the #{arg1} renderer process running your tests just exited unexpectedly.
```
```suggestion
      We detected that the #{arg1} browser process running your tests just exited unexpectedly.
```

flotwig commented on packages/server/lib/errors.coffee:
```diff
@@ -528,23 +528,11 @@ getMsgByType = (type, arg1 = {}, arg2) ->
 
     when "RENDERER_CRASHED"
```
```suggestion
    when "BROWSER_CRASHED"
```

flotwig commented on packages/server/lib/errors.coffee:
```diff
@@ -528,23 +528,11 @@ getMsgByType = (type, arg1 = {}, arg2) ->
-      You can learn more including how to fix Docker here:
-
-      https://on.cypress.io/renderer-process-crashed
+      Learn more at https://on.cypress.io/renderer-process-crashed
```
```suggestion
      Learn more at https://on.cypress.io/browser-process-crashed
```

flotwig commented on packages/server/lib/project.js:
```diff
@@ -368,6 +368,18 @@ class Project extends EE {
+      onUnexpectedDisconnect: () => {
+        const err = errors.get('RENDERER_CRASHED', this.browser.displayName)
+
+        browsers.close(true)
```
just keep it open for debugging

flotwig committed:
changes from review

flotwig committed:
clean up plugins util debug logs

flotwig committed:
clean up s'more debug logs...

flotwig committed:
mo debug logs, fewer problems

flotwig committed:
add basic e2e test

flotwig committed:
update run_spec

flotwig committed:
update how error is thrown

flotwig committed:
fix

flotwig committed:
onceify onUnexpectedDisconnect

flotwig committed:
fix e2e test

flotwig committed:
Merge remote-tracking branch 'origin/develop' into issue-6170-detect-renderer-crashes

flotwig committed:
fix project.js

flotwig committed:
skip searching for a new runner if socket is ended

flotwig committed:
Merge remote-tracking branch 'origin/develop' into issue-6170-detect-renderer-crashes

chrisbreiding commented:
> what do we do when an unexpected exit does occur?

>    browser is left open for debugging
>    in open mode: a warning is logged to stdio

I tried this out and I'm not sure this is the best experience for open mode for the following reasons:

- The browser is completely frozen, so I can't do anything with it. 
- I tend not to look at stdio when in open mode, so I wouldn't have seen the warning if I didn't know it was going to be there. 
- If in global mode, there is no stdio. 

I think it would be better to close the browser and show the error in the desktop GUI.

flotwig commented:
@chrisbreiding It was originally displayed as a warning in the desktop-gui, but since users will see this error frequently (for example, any time they manually exit the browser during tests), it was deemed too noisy.

chrisbreiding commented:
I see. I didn't realize this would happen any time the browser closed.

flotwig committed:
Merge remote-tracking branch 'origin/develop' into issue-6170-detect-renderer-crashes

flotwig commented:
@chrisbreiding with that being said, do you still think it's better to close the browser? The only thing with that is that it could make it difficult to debug the source of websocket disconnections, since the browser would be gone.

chrisbreiding commented:
I didn't realize this was related to websocket disconnects too. I thought it was just the browser freezing or crashing, in which case it seems like the browser isn't accessible for debugging. I guess if there's a way to distinguish, we could handle those cases differently. If not, erring on the side of leaving the browser open is probably preferrable.

brian-mann reviewed (commented):
I'm going to need to pair with you 1:1 before approving this. There are some experience related things I need to verify. We can do this early next week.

flotwig committed:
Merge remote-tracking branch 'origin/develop' into issue-6170-detect-renderer-crashes

flotwig committed:
update error message

flotwig committed:
listenForMochaEndEvent

flotwig committed:
update tests

flotwig committed:
fix video capture when an empty data chunk is returned...

flotwig committed:
pass onError to getElectronProps correctly

flotwig committed:
fix double-logging and logging in run mode

flotwig committed:
only error once per socket

flotwig committed:
no ended

flotwig committed:
don't err if tests ended since d/c

flotwig committed:
ensure 1 error per socket in a cleaner way

flotwig committed:
Merge remote-tracking branch 'origin/develop' into issue-6170-detect-renderer-crashes

flotwig committed:
cleanup

flotwig committed:
only 1 error per start

flotwig committed:
Merge branch 'develop' into issue-6170-detect-renderer-crashes

flotwig committed:
update automation client disconnected text

flotwig committed:
don't use onCrashed as a signal anymore

since we are going to implement proper browser crash detection later, let's use the socket disconnection as the only signal for now

flotwig committed:
Merge branch 'develop' into issue-6170-detect-renderer-crashes

flotwig committed:
add DEBUG logs for cookies_spec

flotwig committed:
Revert "Fix race condition when there's an early asynchronous error in… (#6610)"

This reverts commit 474b80a50f0a8e5e049384b6233ee5f7f3b8dc22.

flotwig committed:
fix lint for app_spec
